src/mbgl/style/style.cpp

```diff
@@ -382,48 +342,111 @@ void Style::update(const UpdateParameters& parameters) {
+    }
+
+    if (lightChanged || zoomChanged || renderLight.hasTransition()) {
+        renderLight.evaluate(evaluationParameters);
+    }
+
+
+    std::vector<Immutable<Source::Impl>> newSourceImpls;
+    newSourceImpls.reserve(sources.size());
+    for (const auto& source : sources) {
+        newSourceImpls.push_back(source->baseImpl);
+    }
+
+    const SourceDifference sourceDiff = diffSources(sourceImpls, newSourceImpls);
+    sourceImpls = std::move(newSourceImpls);
+
+    // Remove render layers for removed sources.
+    for (const auto& entry : sourceDiff.removed) {
+        renderLayers.erase(entry.first);
+    }
+
+    // Create render sources for newly added sources.
+    for (const auto& entry : sourceDiff.added) {
+        std::unique_ptr<RenderSource> renderSource = entry.second->createRenderSource();
+        renderSource->setObserver(this);
+        renderSources.emplace(entry.first, std::move(renderSource));
+    }
+
+    // Update render sources for changed sources.
+    for (const auto& entry : sourceDiff.changed) {
+        renderSources.at(entry.first)->setImpl(entry.second);
+    }
+
 
-    if (cascade) {
-        renderLight->transition(cascadeParameters);
+    std::vector<Immutable<Layer::Impl>> newLayerImpls;
+    newLayerImpls.reserve(layers.size());
+    for (const auto& layer : layers) {
+        newLayerImpls.push_back(layer->baseImpl);
     }
 
-    if (evaluate || renderLight->hasTransition()) {
-        renderLight->evaluate(evaluationParameters);
+    const LayerDifference layerDiff = diffLayers(layerImpls, newLayerImpls);
+    layerImpls = std::move(newLayerImpls);
+
+    // Remove render layers for removed layers.
+    for (const auto& entry : layerDiff.removed) {
+        renderLayers.erase(entry.first);
     }
 
-    for (const auto& renderSource : renderSources) {
-        renderSource->enabled = false;
+    // Create render layers for newly added layers.
+    for (const auto& entry : layerDiff.added) {
+        renderLayers.emplace(entry.first, entry.second->createRenderLayer());
     }
 
-    for (const auto& layer : renderLayers) {
-        if (cascade) {
-            layer->cascade(cascadeParameters);
+    // Update render layers for changed layers.
+    for (const auto& entry : layerDiff.changed) {
+        renderLayers.at(entry.first)->setImpl(entry.second);
+    }
+
+    // Update layers for class and zoom changes.
+    for (const auto& entry : renderLayers) {
+        RenderLayer& layer = *entry.second;
+        const bool layerAdded = layerDiff.added.count(entry.first);
+        const bool layerChanged = layerDiff.changed.count(entry.first);
+
+        if (classesChanged || layerAdded || layerChanged) {
+            layer.cascade(cascadeParameters);
         }
 
-        if (evaluate || layer->hasTransition()) {
-            layer->evaluate(evaluationParameters);
+        if (classesChanged || layerAdded || layerChanged || layer.hasTransition()) {
```

Comment:
`zoomChanged` is missing here, and none of our regression tests caught it. Will investigate.